tests/unit/s2n_self_talk_shutdown_test.c

```diff
@@ -20,67 +20,78 @@ int main(int argc, char **argv)
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+
+    /* Self-Talk: shutdown during application data */
+    for (size_t mode_i = 0; mode_i < s2n_array_len(modes); mode_i++) {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
         EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
-        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
-        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
-
-        /* Verify state prior to alert */
-        EXPECT_FALSE(server_conn->close_notify_received);
-        EXPECT_FALSE(client_conn->close_notify_received);
-
-        /* Verify successful shutdown. */
-        EXPECT_FAILURE_WITH_ERRNO(s2n_shutdown(client_conn, &client_blocked), S2N_ERR_IO_BLOCKED);
-        EXPECT_SUCCESS(s2n_shutdown(server_conn, &server_blocked));
-        EXPECT_SUCCESS(s2n_shutdown(client_conn, &client_blocked));
-
-        /* Verify state after alert */
-        EXPECT_TRUE(server_conn->close_notify_received);
-        EXPECT_TRUE(client_conn->close_notify_received);
-
-        /* Cleanup */
-        EXPECT_SUCCESS(s2n_connection_free(server_conn));
-        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Both the client and server should send application data, but not read any.
+         * This leaves unread application data for both to handle.
+         */
+        uint8_t data[] = "hello world";
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        EXPECT_EQUAL(s2n_send(server_conn, data, sizeof(data), &blocked), sizeof(data));
+        EXPECT_EQUAL(s2n_send(client_conn, data, sizeof(data), &blocked), sizeof(data));
+
+        /* Choose which connection will request the shutdown, and which will respond */
+        struct s2n_connection *request = server_conn;
+        struct s2n_connection *response = client_conn;
+        if (modes[mode_i] == S2N_CLIENT) {
+            request = client_conn;
+            response = server_conn;
+        }
+
+        /* The first shutdown attempt will block on the peer's close_notify */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_shutdown(request, &blocked), S2N_ERR_IO_BLOCKED);
+        EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+        /* The next shutdown attempts should succeed */
```

Comment:
Just to clarify the order of the shutdown here is significant right? The fact that this is response's first shutdown, means the following request's shutdown won't be BLOCKED_ON_READ. 

If the following two lines where in the other order (s.t. shutdown request came first) we would expect a second block on read from request.

Or am I misunderstanding this feature?

Reply:
Updated the comments, see if that is clearer.